lua/nvim_lsp/pyls_ms.lua

```diff
@@ -0,0 +1,133 @@
+local skeleton = require 'nvim_lsp/skeleton'
+local util = require 'nvim_lsp/util'
+local lsp = vim.lsp
+
+local name = "pyls_ms"
+
+local function get_python_version()
+  local f = io.popen("python --version 2>&1") -- runs command
+  local l = f:read("*a") -- read output of command
+  f:close()
+  return l:match("^Python%s*(...).*%s*$")
+end
+
+local function make_installer()
+  local P = util.path.join
+  local install_dir = P{util.base_install_dir, name}
+
+  local bin = P{install_dir, "Microsoft.Python.LanguageServer.dll"}
+  local cmd = {"dotnet", "exec", bin}
+
+  local X = {}
+  function X.install()
+    local install_info = X.info()
+    if install_info.is_installed then
+      print(name, "is already installed")
+      return
+    end
+    if not (util.has_bins("curl")) then
+      error('Need "curl" to install this.')
+      return
+    end
+    if not util.has_bins("sh", "chmod", "unzip") then
+      error('Need the binaries "sh", "chmod", "unzip" to install this')
+      return
+    end
+    local script = [=[
+set -e
+# Install dotnet if not available.
+which dotnet  >/dev/null || {
+  test -x dotnet || {
+    curl -L https://dot.net/v1/dotnet-install.sh | sh
+  }
+  export PATH=$PWD:$PATH
+}
+
+# clone project
+if [ "$(uname)" == "Darwin" ]
```

Comment:
- `vim.fn.has('mac')` (or 'unix' for Unix, or 'win32' for Windows)
- the `curl` steps can be `vim.fn.system('curl ...')`

Then the shell script basically isn't needed. `unzip` isn't available on Windows, but at least if someone comes along wanting to get it working on Windows, there's much less of a wall staring at them :)

Reply:
Ok, I think this is improved in my latest version but I'll need to test on windows. The only thing that I see is guaranteed on (modern) Windows would be powershell's Expand-Archive.

Reply:
> powershell's Expand-Archive.

sure, that's fine. Running powershell scripts can be tricky though, but there's a trick: piping the command through stdin avoids powershell complaints about untrusted scripts: https://github.com/glacambre/firenvim/commit/22db6ad01cb40f2500f85678fa869df095b19c6f#diff-1711f9eb5662d11dbf436e757decb205R284